File: Game/Source/Objects/Player.cpp
#include "GamePCH.h"

#include "Objects/Player.h"
#include "Objects/PlayerController.h"
#include "Game.h"

Player::Player(fw::Scene* pScene, PlayerController* pPlayerController, std::string name, vec3 pos, vec3 rot, vec3 scale, fw::Material* pMaterial)
    : fw::GameObject(pScene, name, pos, rot, scale, pMaterial)
    , m_pPlayerController(pPlayerController)
{
    m_pScene->GetGameCore()->GetEventManager()->RegisterEventListener(this, fw::PhysicsCollisionEvent::GetStaticEventType());
}

Player::Player(Player& original)
    : fw::GameObject(original)
{
    m_pScene->GetGameCore()->GetEventManager()->RegisterEventListener(this, fw::PhysicsCollisionEvent::GetStaticEventType());
    m_pPlayerController = original.m_pPlayerController;
}

Player::~Player()
{

}

void Player::Update(float deltaTime)
{
    GameObject::Update(deltaTime);

    fw::Component* pComponent = GetFirstComponentOfType(fw::CollisionComponent::GetStaticType());
    if (pComponent == nullptr)
    {
        float speed = 2.0f;

        vec2 dir = vec2(0.f, 0.f);

        if (m_pPlayerController->IsHeld(PlayerController::Mask::Up))
        {
            dir.y += 1;
        }
        if (m_pPlayerController->IsHeld(PlayerController::Mask::Down))
        {
            dir.y += -1;
        }
        if (m_pPlayerController->IsHeld(PlayerController::Mask::Left))
        {
            dir.x += -1;
        }
        if (m_pPlayerController->IsHeld(PlayerController::Mask::Right))
        {
            dir.x += 1;
        }

        m_pMaterial->SetUVScale(vec2(64.0 / 1024.0, 64.0 / 512.0));
        m_pMaterial->SetUVOffset(vec2(780.0 / 1024.0, 383.0 / 512.0));

        dir.Normalize();

        m_Position += dir * speed * deltaTime;
    }
    else
    {
        fw::CollisionComponent* pCollider = static_cast<fw::CollisionComponent*>( pComponent );
        fw::PhysicsBody* pBody = pCollider->GetPhysicsBody();

        if (m_pPlayerController->IsHeld(PlayerController::Mask::Right))
        {
            pBody->ApplyForceToCenter(vec3(1, 0, 0) * m_Speed);
        }

        if (m_pPlayerController->IsHeld(PlayerController::Mask::Left))
        {
            pBody->ApplyForceToCenter(vec3(-1, 0, 0) * m_Speed);
        }

        if (m_pPlayerController->WasNewlyPressed(PlayerController::Mask::Up))
        {
            if (m_AllowedToJump)
            {
                pBody->ApplyImpulseToCenter(vec3(0, 1, 0) * m_Speed);
                m_AllowedToJump = false;
            }
        }
    }
}

void Player::OnEvent(fw::Event* pEvent)
{
    if (pEvent->GetType() == fw::PhysicsCollisionEvent::GetStaticEventType())
    {
        m_AllowedToJump = true;
    }
}

void Player::ImGuiInspector()
{
    GameObject::ImGuiInspector();

    ImGui::SliderFloat("Speed", &m_Speed, 0.f, 10.f, "%.2f");
}

File: Game/Source/Objects/Player.lua

-- Player

local dir = vec3(0,0,0)
local bCanJump = true

function Key(button)
	return string.byte(button)
end

function OnInputEvent(go, deviceType, deviceState, key, canJump)
	bCanJump = canJump
	if(deviceType == 0) then -- Is Keyboard
		if(deviceState == 0) then -- Pressed
			if(key == Key("W")) then -- W
				dir.y = 1		
			end
			if(key == Key("A")) then -- A
				dir.x = -1
			end
			if(key == Key("S")) then -- S
				dir.y = -1
			end
			if(key == Key("D")) then -- D
				dir.x = 1
			end
		end
		if(deviceState == 1) then -- Released
			if(key == Key("W")) then -- W
				dir.y = 0
			end
			if(key == Key("A")) then -- A
				dir.x = 0
			end
			if(key == Key("S")) then -- S
				dir.y = 0
			end
			if(key == Key("D")) then -- D
				dir.x = 0
			end
		end
	end
end

function Tick(go, deltaTime)
	local speed = 5
	if(go:GetPhysicsBody() ~= nill) then
		if (dir.x ~= 0) then
			go:GetPhysicsBody():ApplyForceToCenter(dir:Scale(speed))
		end
		if (dir.y > 0 and bCanJump) then
			go:GetPhysicsBody():ApplyImpulseToCenter(dir:Scale(1))
			bCanJump = false -- This is partially working for some reason. TODO: fix it
		end
	else
		dir:Normalize()
		go.position.x = go.position.x + dir.x * deltaTime * speed
		go.position.y = go.position.y + dir.y * deltaTime * speed
	end
end